File: Go/multithread/alternating_print.go
package main

import (
	"fmt"
	"runtime"
	"sync"
	"time"
)

var (
	currentNumber int  = 1
	turnEven      bool = false
	mutex         sync.Mutex
	cond          *sync.Cond
)

const MAX_NUMBER = 1_000_000

func printEven() {
	for currentNumber <= MAX_NUMBER {
		mutex.Lock()
		for !turnEven && currentNumber <= MAX_NUMBER {
			cond.Wait()
		}

		if currentNumber > MAX_NUMBER {
			mutex.Unlock()
			break
		}

		if currentNumber%2 == 0 {
			fmt.Println("Even Thread: ", currentNumber)
			currentNumber++
			turnEven = false
			cond.Signal()
		}

		mutex.Unlock()
	}
}

func printOdd() {
	for currentNumber <= MAX_NUMBER {
		mutex.Lock()
		for turnEven && currentNumber <= MAX_NUMBER {
			cond.Wait()
		}

		if currentNumber > MAX_NUMBER {
			mutex.Unlock()
			break
		}

		if currentNumber%2 == 1 {
			fmt.Println("Odd Thread: ", currentNumber)
			currentNumber++
			turnEven = true
			cond.Signal()
		}

		mutex.Unlock()
	}
}

// 1k Time used: 0.012983 seconds
// 100w Time used: 29.991467 seconds
func main() {
	start := time.Now()

	runtime.GOMAXPROCS(2)

	mutex = sync.Mutex{}
	cond = sync.NewCond(&mutex)

	var wg sync.WaitGroup
	wg.Add(2)

	go func() {
		defer wg.Done()
		printOdd()
	}()

	go func() {
		defer wg.Done()
		printEven()
	}()

	wg.Wait()

	end := time.Now()
	duration := end.Sub(start)
	fmt.Printf("Time used: %.6f seconds\n", duration.Seconds())
}

File: multithread/alternating_print.cc
#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <chrono>
std::mutex mtx;
std::condition_variable cv;
int current_number = 1;
bool even_turn = false;

const int MAX_NUMBER = 1'000;

void print_even() {
    while (current_number <= MAX_NUMBER) {
        std::unique_lock<std::mutex> lock(mtx);
        cv.wait(lock, [] {return even_turn || current_number > MAX_NUMBER;});

        if (current_number > MAX_NUMBER) {
            break;
        }

        if (current_number % 2 == 0) {
            std::cout << "Even Thread: " << current_number << std::endl;
            current_number++;
            even_turn = false;
        }
        
        cv.notify_one();
    }
}

void print_odd() {
    while (current_number <= MAX_NUMBER) {
        std::unique_lock<std::mutex> lock(mtx);
        cv.wait(lock, [] {return !even_turn || current_number > MAX_NUMBER;});

        if (current_number > MAX_NUMBER) {
            break;
        }

        if (current_number % 2 == 1) {
            std::cout << "Odd Thread: " << current_number << std::endl;
            current_number++;
            even_turn = true;
        }

        cv.notify_one();
    }
}

// 1w Time uses: 0.0750917 seconds
// 100w Time uses: 33.8143 seconds
int main() {
    std::chrono::steady_clock::time_point start = std::chrono::steady_clock::now();

    std::thread odd_thread(print_odd);
    std::thread even_thread(print_even);

    odd_thread.join();
    even_thread.join();

    std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
    std::chrono::duration<double> time_used = std::chrono::duration_cast<std::chrono::duration<double>>(end - start);
    std::cout << "Time used: " << time_used.count() << " seconds" << std::endl;

    return 0;
}